File: Tutorial4-2Subcomponents/src/main/java/com/example/tutorial4_2subcomponents/di/ApplicationModule.java
package com.example.tutorial4_2subcomponents.di;

import android.content.Context;
import android.content.SharedPreferences;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class ApplicationModule {

    private Context context;

    public ApplicationModule(Context context) {
        this.context = context;
    }

    @Provides
    @Singleton
    SharedPreferences provideSharedPreferences() {
        return context.getSharedPreferences("PrefName", Context.MODE_PRIVATE);
    }


}

File: Tutorial4-2Subcomponents/src/main/java/com/example/tutorial4_2subcomponents/di/ApplicationModule.kt
package com.example.tutorial4_2subcomponents.di

import android.content.Context
import android.content.SharedPreferences
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
class ApplicationModule(private val context: Context) {

    @Provides
    @Singleton
    fun provideSharedPreferences(): SharedPreferences {
        return context.getSharedPreferences("PrefName", Context.MODE_PRIVATE)
    }

}